File: backend-express/src/models/notifikasiModel.js
import db from "../database/connection.js";

export const findByOrangTua = async (orang_tua_id, page = 1, limit = 20) => {
    const offset = (page - 1) * limit;
    const [rows] = await db.query(
        `SELECT
            n.id,
            n.judul,
            n.pesan,
            n.tipe,
            n.sudah_dibaca,
            n.sent_at,
            n.rujukan_id,
            n.jadwal_id,
            n.pengukuran_id,
            COALESCE(pengukuran_notif.anak_id, pengukuran_rujukan.anak_id) AS anak_id
        FROM notifikasi n
        LEFT JOIN pengukuran pengukuran_notif
            ON pengukuran_notif.id = n.pengukuran_id
        LEFT JOIN rujukan r ON r.id = n.rujukan_id
        LEFT JOIN pengukuran pengukuran_rujukan
            ON pengukuran_rujukan.id = r.pengukuran_id
        WHERE n.orang_tua_id = ?
        ORDER BY n.sent_at DESC
        LIMIT ? OFFSET ?`,
        [orang_tua_id, limit, offset],
    );
    return rows;
};

export const countByOrangTua = async (orang_tua_id) => {
    const [rows] = await db.query(
        `SELECT COUNT(*) AS total
        FROM notifikasi
        WHERE orang_tua_id = ?`,
        [orang_tua_id],
    );
    return parseInt(rows[0].total) || 0;
};

export const countBelumDibaca = async (orang_tua_id) => {
    const [rows] = await db.query(
        `SELECT COUNT(*) AS total
        FROM notifikasi
        WHERE orang_tua_id = ?
        AND sudah_dibaca = FALSE`,
        [orang_tua_id],
    );
    return parseInt(rows[0].total) || 0;
};

export const tandaiDibaca = async (id, orang_tua_id) => {
    const [result] = await db.query(
        `UPDATE notifikasi
        SET sudah_dibaca = TRUE
        WHERE id = ?
        AND orang_tua_id = ?`,
        [id, orang_tua_id],
    );
    return result.affectedRows;
};

export const tandaiSemuaDibaca = async (orang_tua_id) => {
    const [result] = await db.query(
        `UPDATE notifikasi
        SET sudah_dibaca = TRUE
        WHERE orang_tua_id = ?
        AND sudah_dibaca = FALSE`,
        [orang_tua_id],
    );
    return result.affectedRows;
};

export const findById = async (id, orang_tua_id) => {
    const [rows] = await db.query(
        `SELECT id FROM notifikasi
        WHERE id = ? AND orang_tua_id = ?`,
        [id, orang_tua_id],
    );
    return rows[0] || null;
};

export const deleteById = async (id, orang_tua_id) => {
    const [result] = await db.query(
        `DELETE FROM notifikasi
        WHERE id = ? AND orang_tua_id = ?`,
        [id, orang_tua_id],
    );
    return result.affectedRows;
};

export const deleteSudahDibaca = async (orang_tua_id) => {
    const [result] = await db.query(
        `DELETE FROM notifikasi
        WHERE orang_tua_id = ?
        AND sudah_dibaca = TRUE`,
        [orang_tua_id],
    );
    return result.affectedRows;
};

File: tumbuhapp/lib/shared/models/notifikasi_model.dart
class NotifikasiModel {
  final int id;
  final String judul;
  final String pesan;
  final String tipe;
  final bool sudahDibaca;
  final String sentAt;
  final int? rujukanId;
  final int? jadwalId;
  final int? pengukuranId;
  final String? anakId;

  NotifikasiModel({
    required this.id,
    required this.judul,
    required this.pesan,
    required this.tipe,
    required this.sudahDibaca,
    required this.sentAt,
    this.rujukanId,
    this.jadwalId,
    this.pengukuranId,
    this.anakId,
  });

  factory NotifikasiModel.fromJson(Map<String, dynamic> json) {
    return NotifikasiModel(
      id: json['id'],
      judul: json['judul'] ?? '',
      pesan: json['pesan'] ?? '',
      tipe: json['tipe'] ?? '',
      sudahDibaca: json['sudah_dibaca'] == 1 || json['sudah_dibaca'] == true,
      sentAt: json['sent_at'] ?? '',
      rujukanId: _asInt(json['rujukan_id']),
      jadwalId: _asInt(json['jadwal_id']),
      pengukuranId: _asInt(json['pengukuran_id']),
      anakId: json['anak_id']?.toString(),
    );
  }

  NotifikasiModel copyWith({bool? sudahDibaca}) {
    return NotifikasiModel(
      id: id,
      judul: judul,
      pesan: pesan,
      tipe: tipe,
      sudahDibaca: sudahDibaca ?? this.sudahDibaca,
      sentAt: sentAt,
      rujukanId: rujukanId,
      jadwalId: jadwalId,
      pengukuranId: pengukuranId,
      anakId: anakId,
    );
  }

  static int? _asInt(dynamic value) {
    if (value == null) return null;
    if (value is num) return value.toInt();
    return int.tryParse(value.toString());
  }
}
